File: q3/q3b-join-impl.c
#include <stdio.h>
#include <assert.h>
#include <pthread.h>

void *request (void *args) {
    int random_number = (rand() % 30) + 1; // // Obtain a number between [1 - 30].
    sleep(random_number); // Sleeps seconds
    pthread_exit(random_number);
}

int gateway (int num_replicas) {
    pthread_t pthreads[num_replicas];
    for (int i = 0; i < num_replicas; i++) {
        pthread_create (&pthreads[i], NULL, &request, (void*) i);
    }
    int sum = 0;
    for (int i = 0; i < num_replicas; i++) {
        int aux;
        pthread_join(pthreads[i], &aux);
        sum += aux;
    }
    return sum;
}

int main (int argc, char *argv[]) {
    printf("gateway=%d\n", gateway(5));
}

File: q3/q3bJoinImpl.java
import java.util.Random;

public class Main {

    public static void main(String args[]) {
        System.out.println("gateway=" + gateway(5));
    }

    public static int gateway(int numReplicas) {
        Sum sum = new Sum();
        Request request = new Request(sum);
        Thread[] threads = new Thread[numReplicas];
        for (int i = 0; i < numReplicas; i++) {
            Thread thread = new Thread(request, "request" + i);
            thread.start();
            threads[i] = thread;
        }

        try {
            System.out.println("Joiner waititng...");
            for (int i = 0; i < numReplicas; i++) {
                threads[i].join();
                System.out.printf("Thread %s finished.\n", threads[i].getName());
            }
        } catch (InterruptedException e) {}
        return sum.get();
    }

    public static class Request implements Runnable {
        private Sum sum;

        public Request (Sum sum) {
            this.sum = sum;
        }

        public void run() {
            int randomNumber = (new Random()).nextInt(30) + 1; // Obtain a number between [1 - 30].
            try {                
                System.out.printf("Request will sleep %d seconds.\n", randomNumber);
                Thread.sleep(randomNumber * 1000); // Thread.sleep sleeps milliseconds
            } catch (InterruptedException e) {}
            this.sum.add(randomNumber);
        }
    }

    public static class Sum {
        private int value;

        public synchronized void add(int value) {
            this.value += value;
        }

        public synchronized int get() {
            return this.value;
        }
    }  

}

